Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XLGPlusPlus/XldbProtobufExtensions.cs

```diff
@@ -558,5 +548,479 @@ public static Xldb.Fingerprint ToFingerprint(this Fingerprint fingerprint)
+
+        /// <nodoc />
+        public static Xldb.PipProvenance ToPipProvenance(this PipProvenance provenance)
+        {
+            return provenance == null ? null : new Xldb.PipProvenance()
+            {
+                Usage = provenance.Usage.ToPipData(),
+                ModuleId = provenance.ModuleId.Value.Value,
+                ModuleName = provenance.ModuleName.ToString(),
+                SemiStableHash = provenance.SemiStableHash
+            };
+        }
+
+        public static Xldb.FileOrDirectoryArtifact ToFileOrDirectoryArtifact(this FileOrDirectoryArtifact artifact, PathTable pathTable)
+        {
+            var xldbFileOrDirectoryArtifact = new Xldb.FileOrDirectoryArtifact()
+            {
+                IsValid = artifact.IsValid
+            };
+
+            if (artifact.IsDirectory)
+            {
+                xldbFileOrDirectoryArtifact.IsDirectory = true;
+                xldbFileOrDirectoryArtifact.DirectoryArtifact = artifact.DirectoryArtifact.ToDirectoryArtifact(pathTable);
+            }
+            else
+            {
+                xldbFileOrDirectoryArtifact.IsFile = true;
+                xldbFileOrDirectoryArtifact.FileArtifact = artifact.FileArtifact.ToFileArtifact(pathTable);
+            }
+
+            return xldbFileOrDirectoryArtifact;
+        }
+
+        /// <nodoc />
+        public static Xldb.Pip ToPip(this Pip pip, DirectedGraph directedGraph)
+        {
+            var xldbPip = new Xldb.Pip()
+            {
+                SemiStableHash = pip.SemiStableHash,
+                FormattedSemiStableHash = pip.FormattedSemiStableHash,
+                ProcessAllowsUndeclaredSourceReads = pip.ProcessAllowsUndeclaredSourceReads,
+                PipId = pip.PipId.Value,
+                PipType = (PipType)pip.PipType,
+                Provenance = pip.Provenance.ToPipProvenance(),
+            };
+
+            if (pip.Tags.IsValid)
+            {
+                xldbPip.Tags.AddRange(pip.Tags.Select(key => key.ToString()));
+            }
+
+            xldbPip.IncomingEdges.AddRange(directedGraph.GetIncomingEdges(pip.PipId.ToNodeId()).Select(edge => edge.ToEdge()));
+            xldbPip.OutgoingEdges.AddRange(directedGraph.GetOutgoingEdges(pip.PipId.ToNodeId()).Select(edge => edge.ToEdge()));
+            return xldbPip;
+        }
+
+        /// <nodoc />
+        public static Xldb.ModulePip ToModulePip(this ModulePip pip, PathTable pathTable, Xldb.Pip parentPip)
+        {
+            var xldbModulePip = new Xldb.ModulePip
+            {
+                ParentPipInfo = parentPip,
+                Module = pip.Module.Value.Value,
+                Identity = pip.Identity.ToString(),
+                ResolverKind = pip.ResolverKind.ToString(),
+                ResolverName = pip.ResolverName.ToString(),
+                Version = pip.Version.ToString(),
+                Location = new Xldb.LocationData()
+                {
+                    IsValid = pip.Location.IsValid,

```

Comment:
>IsValid [](start = 20, length = 7)

If protobuf generates classes we shouldn't need these IsValid bools to be serialized since we can just use null for unpopulated references

Reply:
Actually you are right. Wherever there is invalid, I can just not generate that class. Let me go ahead and change those.

---
In reply to: [314454672](https://github.com/microsoft/BuildXL/pull/724#discussion_r314454672) [](ancestors = 314454672)